lib/luacrypto_aead.c

```diff
@@ -0,0 +1,325 @@
+	luaL_pushresult(&b);
+	lunatik_free(iv_data);
+	return 1;
+}
+
+/***
+* Decrypts data using the AEAD transform.
+* The IV (nonce) and AAD must match those used during encryption.
+* @function crypto_aead:decrypt
+* @tparam string iv The Initialization Vector (nonce). Its length must match `ivsize()`.
+* @tparam string combined_data A string containing AAD (Additional Authenticated Data) concatenated with the ciphertext and tag (format: AAD || Ciphertext || Tag).
+* @tparam integer aad_len The length of the AAD part in `combined_data`.
+* @treturn string The decrypted data, formatted as (AAD || Plaintext).
+* @raise Error on decryption failure (e.g., authentication error - EBADMSG), incorrect IV length, input data too short, or allocation issues.
+*/
+static int luacrypto_aead_decrypt(lua_State *L) {
+	struct crypto_aead *tfm = luacrypto_aead_check(L, 1);
+	size_t iv_len, combined_len;
+	const char *iv = luaL_checklstring(L, 2, &iv_len); /* Nonce */
+	const char *combined = luaL_checklstring(L, 3, &combined_len);
+	lua_Integer aad_l = luaL_checkinteger(L, 4);
+
+	u8 *iv_data = luacrypto_aead_prepare_iv(L, tfm, iv, iv_len);
+
+	luaL_argcheck(L, aad_l >= 0 && (size_t)aad_l <= combined_len, 4, "AAD length out of bounds");
+	size_t aad_len = (size_t)aad_l;
+	size_t ciphertext_with_tag_len = combined_len - aad_len; /* CiphertextWithTag length */
+
+	unsigned int authsize_val = crypto_aead_authsize(tfm);
+	/*
+	* For decryption, buf needs to hold AAD || CiphertextWithTag initially.
+	* The output (AAD || Plaintext) will be shorter or same length.
+	* Kernel expects source buffer to be large enough for AAD + CT + Tag.
+	* Destination (which is same buffer) will receive AAD + PT.
+	* So, initial combined_len is AAD + CT + Tag.
+	*/
+
+	luaL_Buffer b;
+	luaL_buffinit(L, &b);
+	char *buf = luaL_prepbuffsize(&b, combined_len);
+	memcpy(buf, combined, combined_len); /* Copies AAD || CiphertextWithTag */
+	luaL_addsize(&b, combined_len);
+
+	/* Scatterlist covers AAD and CiphertextWithTag. */
+	struct scatterlist sg_work;
+	sg_init_one(&sg_work, buf, aad_len + ciphertext_with_tag_len);
+
+	struct aead_request *req = luacrypto_aead_prepare_request(L, tfm, aad_len, ciphertext_with_tag_len, &sg_work, iv_data);
+	if (!req) {
+		lunatik_free(iv_data);
+		lua_error(L);
+	}
+
+	int ret = crypto_aead_decrypt(req);
+	if (ret < 0) {
+		lunatik_free(iv_data);
+		luaL_error(L, "AEAD decryption failed (err %d)", ret);
+	}
+	aead_request_free(req);
+
+	/*
+	* After decryption, buf contains: AAD || Plaintext.
+	* The length of Plaintext is (ciphertext_with_tag_len - authsize_val)
+	*/
+	luaL_argcheck(L, ciphertext_with_tag_len >= authsize_val, 3,
+		"input data (ciphertext+tag) too short for tag"
+	);
+	/* The length of this combined data is (aad_len + plaintext_len) */
+	luaL_addsize(&b, -(int)authsize_val); /* Remove the tag length from the buffer size */
+	luaL_pushresult(&b);
+	lunatik_free(iv_data);
+	return 1;
+}
```

Comment:
why not using the macro template for encrypt/decrypt which we have discussed in the other PR.. my impression is that we could apply previous discussions to make this module more concise..